### ISSUE TITLE
Collection update can return documents different then "Sucessfully Updated Records"

### ISSUE THREAD

**Issue by robertbak** [balderdashy/sails-mongo#348]
This affects queries where fields that are being updated by one of the queries are in search criteria for another concurrent query. This can lead to non-updated records being returned as updated or not returning updated records.

------------------------------------
### Thought experiment:
Imagine two updates, exactly the same, being run one after the other - let's call them A and B, updating a field that's in search criteria.
 
Looking into the update code here (https://github.com/balderdashy/sails-mongo/blob/master/lib/collection.js#L219) and because it's async what can happen is:

A: ```collection.find(query.criteria.where)``` - let's say it returns 10 items it thinks it will update
B: ```collection.find(query.criteria.where)``` - also returns 10 items 
A: ```collection.update(query.criteria.where,...)``` - updates 10 items
B: ```collection.update(query.criteria.where,...)``` - updates 0 items

But both will return all 10 items as if they both updated them. You can get different wrong results by using different queries, probably the worse result being items being updated without being returned.

------------------------------------

One possible (simple) solution to this would be using the found id's for the ```collection.update``` search criteria, but that has the downside of possibly updating fields on items that no longer match the search criteria. 

### Single query working wrong
Another example of this not working correctly is time based queries. Let's say you want to mark records as expired using a date field,  if the ```find``` call takes Xms the update can expire items that are Xms older while never giving the sails app a way to react to the new expiration.

**Comment by sailsbot** [comment]
Thanks for posting, @robertbak.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
Reopening this for discussion.  @robertbak do you think this is a bug with a valid solution that we could put into place?

**Comment by robertbak** [comment]
It's a bit tricky since, as far as I know, with mongo there is no way to get the updated records. So the way it's done now it sometimes returns things which weren't updated or doesn't return ones that were updated  (but this will mostly happen if the update and search queries overlap). The only alternative I could think of is using the _id_ of the documents found in the first query and then you return the correct documents (the ones which were updated) but the update can go through on things which no longer should be updated. 

As it is now, it seems the bug is in the over-promising of the documentation. So there's probably no way to get it exactly right (without crazy things like killing concurrency of updates or running the update element by element), so maybe the solution is to add an asterisk in the documentation for mongo (and other adapters possibly, haven't looked into this).

**Comment by sailsbot** [comment]
Thanks for posting, @robertbak.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
